### PR TITLE
Rework of RSSI% for CRSF/ExpLRS

### PR DESCRIPTION
**Pull request by OptimusTi**
Currently, INAV is using a direct linear conversion to show RSSI in dBm as percentage values.
Since RF signals do not attenuate in a linear way, it makes the current model impractical.

This PR uses a quadratic model that provides a more consistent mapping between dBm and percentages
providing a better representation in percentage of expected signal strength.

RX sensitivity can be adjusted to "tune" the RSSI if needed. Great for ExpressLRS different RF Modes.

img1

### CODE REVIEW

**Comment by avsaase** [comment]
Can you elaborate on why you've chosen this curve and not another one? 

I played around with this a bit a few months ago and I used the rule that with 6dBm above the minimum sensitivity you can go twice as far as you are now. This resulted in a curve that stuck to 100% for much longer.

File: src/main/rx/crsf.c [hunk 1]
```diff
@@ -241,8 +242,15 @@ STATIC_UNIT_TESTED uint8_t crsfFrameStatus(rxRuntimeConfig_t *rxRuntimeConfig)
             rxLinkStatistics.uplinkTXPower = crsfPowerStates[linkStats->uplinkTXPower];
             rxLinkStatistics.activeAnt = linkStats->activeAntenna;
 
-            if (rxLinkStatistics.uplinkLQ > 0)
-                lqTrackerSet(rxRuntimeConfig->lqTracker, scaleRange(constrain(rxLinkStatistics.uplinkRSSI, -120, -30), -120, -30, 0, RSSI_MAX_VALUE));
+            if (rxLinkStatistics.uplinkLQ > 0) {
+                int16_t uplinkStrength;   // RSSI dBm converted to %
+                uplinkStrength = constrain((100 * ((osdConfig()->rssi_dbm_max - osdConfig()->rssi_dbm_min) * (osdConfig()->rssi_dbm_max  - osdConfig()->rssi_dbm_min)) - (osdConfig()->rssi_dbm_max  - rxLinkStatistics.uplinkRSSI) * (100 * (osdConfig()->rssi_dbm_max  - rxLinkStatistics.uplinkRSSI))) / ((osdConfig()->rssi_dbm_max  - osdConfig()->rssi_dbm_min) * (osdConfig()->rssi_dbm_max  - osdConfig()->rssi_dbm_min)),0,100);
```
Comment (avsaase) [inline src/main/rx/crsf.c]: If you put this line in the if/else construct below it doesn't get calculated when RSSI is outside the 0-100% range.

Comment (avsaase) [reply src/main/rx/crsf.c]: Also, this line can be simplified to 
```
uplinkStrength = constrain((100 - (100 * sq(osdConfig()->rssi_dbm_max  - osdConfig()->rssi_dbm_min) / sq(osdConfig()->rssi_dbm_max - rxLinkStatistics.uplinkRSSI))), 0, 100);
```

**Comment by avsaase** [comment]
Here's my attempt at making a plot with the proposed defaults from the PR. I tried to find some references for the equation you used but was not able to find them. 

I also added the scaling ArduPilot uses for CRSF (linear between -50 and -120 dBm) and my own distance-based scaling. The advantage of the latter is that it is just based on the attenuation of the signal with increasing distance. When you are at 50% RSSI you know you are at half your total range. You also don't need to specify a dBm value where the signal is "good". 

You could argue that a 6 dBm link margin does not really mean you can go twice as far as your current distance, but we could easily include a real-world penalty to make it more realistic.

file_show

The formula for the distance-based RSSI(%) is 
```
linkMargin = currentRssi - minSensitivity
timesFurther = 10^(linkMargin / (20 * realWorldPenaltyFactor)) // realWorldPenaltyFactor = 1.5 for the "realistic" curve
RSSI(%) = 100 - 100 / timesFurther
```

File: src/main/rx/crsf.c (hunk 1 above)
Comment (OptimusTi) [reply src/main/rx/crsf.c]: Thanks for the input. I will update with sq() but there's no need to square everything

**Comment by OptimusTi** [comment]
My PR was inspired by this article and the formula used for the Linux ipw2x00 driver. 

https://www.intuitibits.com/2016/03/23/dbm-to-percent-conversion/

https://github.com/torvalds/linux/blob/9ff9b0d392ea08090cd1780fb196f36dbb586529/drivers/net/wireless/intel/ipw2x00/ipw2200.c#L4321

The PR is to convert dBm into a percentage value. That's all. RSSI in dBm  comes directly from the RF chip. There no need to complicate things.

File: src/main/rx/crsf.c (hunk 1 above)
Comment (avsaase) [reply src/main/rx/crsf.c]: I only simplified the calculation by canceling terms.

**Comment by avsaase** [comment]
> The PR is to convert dBm into a percentage value. That's all. RSSI in dBm comes directly from the RF chip. There no need to complicate things.

As that article already points out:
> when we say that a network has 75% signal strength, what does that mean?

Nothing! Unless you base the mapping from dBm to % on a physical process.

**Comment by OptimusTi** [comment]
> > The PR is to convert dBm into a percentage value. That's all. RSSI in dBm comes directly from the RF chip. There no need to complicate things.
> 
> As that article already points out:
> 
> > when we say that a network has 75% signal strength, what does that mean?
> 
> Nothing! Unless you base the mapping from dBm to % on a physical process.

please read the whole article.

**Comment by avsaase** [comment]
I did and it doesn't give any substantiation for the equation it uses (same for the ipw2x00 driver). It just mentions the non-linearity but it doesn't address the type of non-linearity in any way.

My point is that any conversion is totally arbitrary when not based on the physical attenuation process. The article even says
> The model’s parameters have been adjusted based on my own observations, and although we can still argue that the accuracy of the model might not be the best, I believe it is more consistent than the lookup table and better than a linear conversion, especially if you again ask yourself the question what does 75% signal strength mean, not only from a user’s perspective but also how it relates to the original dBm measurement.

To me, that just means "I pulled these numbers out of thin air but they _feel_ roughly correct".

**Comment by OptimusTi** [comment]
Thank you for your input. From what I see in the original driver code, the formula has been used for many years and it's still valid.

I'm confident that Mr. Granados and Intel didn't use numbers "out of the air". Anyway, there's no need for the attitude. I'm just trying to help. If someone else with actual RF knowledge could jump in that would be great.

**Review by avsaase — APPROVED** [review]

**Pull request merged**